### PR TITLE
JAMES-3432 Upload: Attachment

### PR DESCRIPTION
**Pull request by nvduc91**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/UploadRoutes.scala [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+  def fromAttachment(attachmentMetadata: AttachmentMetadata): UploadResponse =
+    UploadResponse(
+        blobId = BlobId.of(attachmentMetadata.getAttachmentId.getId).get,
+        `type` = ContentType.of(attachmentMetadata.getType.asString),
+        size = sanitizeSize(attachmentMetadata.getSize),
+        expires = None)
+}
+
+case class UploadResponse(blobId: BlobId,
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/UploadRoutes.scala:75]: Upliad response needs to have an accountId field. Why did you add an expires field?

**Review by chibenwa — COMMENTED** [review]
Looks good :-)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/UploadContract.scala [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+      .basePath("")
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+    .when
+      .get(s"/download/$ALICE_ACCOUNT_ID/$blobId")
+    .`then`
+      .statusCode(SC_UNAUTHORIZED)
+  }
+
+  @Test
+  @Disabled("Disabled by JAMES-1788")
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/UploadContract.scala:131]: `@Disabled("JAMES-1788 Upload size limitation needs to be contributed")`

And add a comment in the issue also.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/UploadRoutes.scala [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                          `type`: ContentType,
+                          size: Size)
+
+class UploadRoutes @Inject()(@Named(InjectionKeys.RFC_8621) val authenticator: Authenticator,
+                             val attachmentManager: AttachmentManager,
+                             val serializer: UploadSerializer) extends JMAPRoutes {
+
+  class CancelledUploadException extends RuntimeException {
+
+  }
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/UploadRoutes.scala:81]: Is that empty body needed ?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/UploadRoutes.scala [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        val targetAccountId: AccountId = AccountId(id)
+        AccountId.from(session.getUser).map(accountId => accountId.equals(targetAccountId))
+          .fold[SMono[Void]](
+            e => SMono.raiseError(e),
+            value => if (value) {
+              SMono.fromCallable(() => ReactorUtils.toInputStream(request.receive.asByteBuffer))
+              .flatMap(content => handle(targetAccountId, contentType, content, session, response))
+              .subscribeOn(Schedulers.elastic())
+            } else {
+              SMono.raiseError(new UnauthorizedException("You cannot upload to others"))
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/UploadRoutes.scala:122]: Attempt to upload in another account

**Review by chibenwa — COMMENTED** [review]
Looks really good!

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged
